stan/math/rev/arr/functor/map_rect_mpi.hpp

```diff
@@ -0,0 +1,350 @@
+
+          // checks if the data is already cached
+          distribute_map_rect_data(x_r, x_i, uid);
+
+          //std::cout << "root uses UID = " << uid_ << std::endl;
+
+          // make childs aware of upcoming job
+          mpi_broadcast_command<stan::math::mpi_distributed_apply<distributed_map_rect<F> > >();
+
+          //std::cout << "setting up root with uid = " << uid << std::endl;
+          setup(uid);
+          
+          // sent uid and # of params
+          std::vector<int> meta(3, 0);
+          meta[0] = uid;
+          meta[1] = E_ ;
+          meta[2] = T_ ;
+          boost::mpi::broadcast(world_, meta.data(), 3, 0);
+          
+          distribute_param(eta, theta);
+
+          (*this)();
+        }
+
+        distributed_map_rect() {
+          //std::cout << "Setting up distributed map on worker " << world_.rank() << " / " << W_ << std::endl;
+          if(R_ == 0)
+            throw std::runtime_error("problem sizes must be defined on the root.");
+
+          //std::cout << "setting up child ..." << std::endl;
+          
+          // get uid & # of params from root
+          std::vector<int> meta(3);
+          boost::mpi::broadcast(world_, meta.data(), 3, 0);
+          uid_ = meta[0];
+          E_ = meta[1];
+          T_ = meta[2];
+
+          //std::cout << "worker " << world_.rank() << " / " << W_ << " uses UID = " << uid_ << std::endl;
+
+          setup(uid_);
+          
+          distribute_param(std::vector<var>(), std::vector<std::vector<var> >());
+
+          (*this)();
+        }
+        
+        static void distributed_apply() {
+          // entry point when called on remote
+
+          // call default constructor
+          distributed_map_rect<F> job_chunk;
+        }
+
+        // process work chunks; collect results at root
+        void operator()() {
+          using std::vector;
+
+          const distributed_map_rect_data& local_ = distributed_data::get_const_instance().find(uid_)->second;
+          
+          vector<double> local_result;
+          // reserve output size for 5 function outputs per job... we
+          // will find out during evaluation
+          local_result.reserve(C_*5*(1+E_+T_));
+
+          // number of outputs per job
+          vector<int> local_F_out(C_, 0);
+
+          try {
+            vector<double>::const_iterator local_theta_iter = local_theta_.begin();
+            vector<double> grad(E_+T_);
+            //for(std::size_t m=0; m != eta_run_v.size(); m++)
+            //  std::cout << "eta_run_v = " << value_of(eta_run_v[m]) << std::endl;
+            for(std::size_t i = 0; i != C_; i++) {
+              start_nested();
+              //std::cout << "job i = " << i << std::endl;
+              
+              // I am not sure why the eta_v must be declared inside
+              // the nested block? I think it should be possible to
+              // have this outside of the for loop, but this gives
+              // wrong results
```

Comment:
I think when you declare something with vars in it and initialize it, it puts things on the global autodiff stack. Then calling start_nested puts that stack aside and begins a new one.

Reply:
Yeah, I figured that the hard way. I am only concerned about performance; I hope the way I put it is good.